File: buble sort/time_merge_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
#include <cstdlib>
#include <fstream>


using namespace std;
void export_to_csv_cpp(const vector<pair<int, double>>& results) {
    ofstream csvfile("results_cpp.csv");

    csvfile << "Tamaño de la lista, Tiempo de ejecución (segundos)" << endl;

    for (const auto& result : results) {
        csvfile << result.first << ", " << result.second << endl;
    }

    csvfile.close();
}

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; ++i) {
        for (int j = 0; j < n - i - 1; ++j) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

double measureTimeSorting(void (*algorithm)(vector<int>&), int listSize) {
    vector<int> arr(listSize);
    srand(time(0));
    for (int i = 0; i < listSize; ++i) {
        arr[i] = rand() % 100000 + 1;
    }

    clock_t start_time = clock();
    algorithm(arr);
    clock_t end_time = clock();

    return static_cast<double>(end_time - start_time) / CLOCKS_PER_SEC;
}

int main() {
    vector<int> listSizes = {100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000};

    cout << "Tiempos de ejecución para diferentes tamaños de entrada:" << endl;
    cout << "| Tamaño de la lista | Tiempo de ejecución (segundos) |" << endl;
    cout << "|--------------------|-------------------------------|" << endl;

    vector<pair<int, double>> results;

    for (int size : listSizes) {
        double executionTime = measureTimeSorting(bubbleSort, size);
        results.push_back(make_pair(size, executionTime));
        cout << "| " << size << "\t\t  | " << executionTime << "\t\t   |" << endl;
    }

    cout << "Resultados almacenados en el vector 'results':" << endl;
    for (const auto& result : results) {
        cout << result.first << " - " << result.second << endl;
    }


    return 0;
}

File: practica_01/merge sort/time_merge_sort.go
package main

import (
	"encoding/csv"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

func merge(arr []int, left int, mid int, right int) {
	leftSize := mid - left + 1
	rightSize := right - mid

	leftHalf := make([]int, leftSize)
	rightHalf := make([]int, rightSize)

	copy(leftHalf, arr[left:left+leftSize])
	copy(rightHalf, arr[mid+1:mid+1+rightSize])

	i, j, k := 0, 0, left

	for i < leftSize && j < rightSize {
		if leftHalf[i] <= rightHalf[j] {
			arr[k] = leftHalf[i]
			i++
		} else {
			arr[k] = rightHalf[j]
			j++
		}
		k++
	}

	for i < leftSize {
		arr[k] = leftHalf[i]
		i++
		k++
	}

	for j < rightSize {
		arr[k] = rightHalf[j]
		j++
		k++
	}
}

func mergeSort(arr []int, left int, right int) {
	if left >= right {
		return
	}

	mid := left + (right-left)/2
	mergeSort(arr, left, mid)
	mergeSort(arr, mid+1, right)
	merge(arr, left, mid, right)
}

func measureTimeSorting(listSize int) time.Duration {
	arr := make([]int, listSize)
	rand.Seed(time.Now().UnixNano())

	for i := 0; i < listSize; i++ {
		arr[i] = rand.Intn(100000) + 1
	}

	startTime := time.Now()
	mergeSort(arr, 0, len(arr)-1)
	endTime := time.Now()

	return endTime.Sub(startTime)
}

func exportToCSVGo(results [][]string) error {
	file, err := os.Create("results_go.csv")
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	for _, result := range results {
		err := writer.Write(result)
		if err != nil {
			return err
		}
	}

	return nil
}

func main() {
	listSizes := []int{100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000}

	fmt.Println("Tiempos de ejecución para diferentes tamaños de entrada:")
	fmt.Println("| Tamaño de la lista | Tiempo de ejecución (segundos) |")
	fmt.Println("|--------------------|-------------------------------|")
	var results [][]string
	for _, size := range listSizes {
		executionTime := measureTimeSorting(size)
		fmt.Printf("| %6d\t\t  | %.6f\t\t   |\n", size, executionTime.Seconds())
		sizeStr := strconv.Itoa(size)
		timeStr := fmt.Sprintf("%.6f", executionTime.Seconds())
		results = append(results, []string{sizeStr, timeStr})
	}

	fmt.Println("Resultados almacenados en el vector 'results':")
	fmt.Println(results)

	err := exportToCSVGo(results)
	if err != nil {
		fmt.Println("Error al exportar a CSV:", err)
	}
}
